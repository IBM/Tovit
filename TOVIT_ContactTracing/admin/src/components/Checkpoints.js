/* globals alert, FileReader */
import React from 'react'
import { useTranslation } from 'react-i18next'
import { postCheckpoints } from '../helpers/api'
import { isValidJSON } from '../helpers/json'

export function Checkpoints ({ onUpload }) {
  const { t } = useTranslation()
  const webAppUrl = process.env.REACT_APP_WEB_APP_DOMAIN

  const handleUpload = () => {
    const files = document.getElementById('checkpoint-upload').files
    if (files.length <= 0) {
      return false
    }
    const fr = new FileReader()
    fr.onload = async (e) => {
      if (e && e.target.result && isValidJSON(e.target.result)) {
        const checkpointsJSON = JSON.parse(e.target.result)
        const uploadSuccess = await postCheckpoints(checkpointsJSON)
        if (uploadSuccess) {
          alert(t('upload_success'))
        } else {
          alert(t('upload_fail'))
        }
        document.getElementById('checkpoint-upload-form').reset()
      } else {
        alert(t('upload_fail'))
      }
    }
    fr.readAsText(files.item(0))
  }

  return (
    <div>
      <div class='d-flex justify-content-between flex-wrap flex-md-nowrap align-items-center pt-3 pb-2 mb-3 border-bottom'>
        <h1 class='h2'>{t('upload_title')}</h1>
      </div>
      <form id='checkpoint-upload-form'>
        <p>{t('upload_instructions_1')} {webAppUrl}. {t('upload_instructions_2')}</p>
        <p>{t('upload_instructions_3')}</p>
        <div class='alert alert-success' style={{ display: 'none' }} id='upload-success' role='alert'>
          {t('upload_success')}
        </div>
        <input class='btn btn-large btn-warning' type='file' id='checkpoint-upload' onChange={handleUpload} />
      </form>
    </div>
  )
}
