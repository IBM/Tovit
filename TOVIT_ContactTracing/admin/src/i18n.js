import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import en from './translations/en'

// the translations
// (tip move them in a JSON file and import them)
const resources = {
  en
}

i18n
  .use(LanguageDetector)
  .use(initReactI18next) // passes i18n down to react-i18next
  .init({
    resources,
    fallbackLng: 'en',
    keySeparator: false, // we do not use keys in form messages.welcome
    interpolation: {
      escapeValue: false // react already safes from xss
    },
    detection: {
      order: ['querystring', 'cookie', 'localStorage', 'navigator', 'htmlTag', 'path', 'subdomain'],

      // keys or params to lookup language from
      lookupQuerystring: 'lng',
      lookupCookie: 'i18next',
      lookupLocalStorage: 'i18nextLng',
      lookupFromPathIndex: 0,
      lookupFromSubdomainIndex: 0,

      // cache user language on
      caches: ['localStorage', 'cookie'],
      excludeCacheFor: ['cimode'], // languages to not persist (cookie, localStorage)

      // optional expire and domain for set cookie
      cookieMinutes: 10,
      cookieDomain: 'myDomain',

      // optional htmlTag with lang attribute, the default is:
      htmlTag: document.documentElement,

      // only detect languages that are in the whitelist
      checkWhitelist: true
    }
  })

export default i18n
