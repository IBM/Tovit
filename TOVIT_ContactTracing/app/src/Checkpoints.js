import React from 'react'
import Grid from '@material-ui/core/Grid'
import Button from '@material-ui/core/Button'
import Typography from '@material-ui/core/Typography'
import CropFreeIcon from '@material-ui/icons/CropFree'
import HomeIcon from '@material-ui/icons/Home'
import ArrowBackIcon from '@material-ui/icons/ArrowBack'
import StopIcon from '@material-ui/icons/Stop'
import QRCode from 'qrcode.react'
import QRReader from 'react-qr-reader'
import { Translation } from 'react-i18next'
import API from './api'

const checkpointKeyLength = Number(process.env.REACT_APP_CHECKPOINT_KEY_LENGTH)

const initialState = {
  mode: 'home',
  checkpointKey: null,
  checkpointTime: null,
  legacyMode: false
}

class Checkpoints extends React.Component {
  constructor () {
    super()
    this.state = initialState
  }

  async reset () {
    this.setState(initialState)
    this.props.resetUrlScanState()
  }

  async becomeHost () {
    try {
      const { key, timestamp } = await API.hostCheckpoint()
      this.setState({
        mode: 'host',
        checkpointKey: key,
        checkpointTime: timestamp
      })
    } catch (e) {
      console.error(e)
      window.alert('There was an unexpected error. Please leave feedback so the developer can fix this.')
    }
  }

  async endHost () {
    // TODO confirmation
    this.reset()
  }

  async joinCheckpoint () {
    this.setState({
      mode: 'join'
    })
  }

  async handleScan (data) {
    if (data) {
      if (data.length === checkpointKeyLength) {
        try {
          await API.joinCheckpoint(data)
          this.setState({ mode: 'scan-success' })
        } catch (e) {
          console.error(e)
          this.setState({ mode: 'scan-error' })
        }
      } else {
        // QR code may be a url
        const urlSplit = data.split('?checkpoint=')
        if ((urlSplit.length === 2) && (urlSplit[1].length === checkpointKeyLength)) {
          await API.joinCheckpoint(urlSplit[1])
          this.setState({ mode: 'scan-success' })
        } else {
          this.setState({ mode: 'scan-error' })
        }
      }
    }
  }

  handleScanError () {
    this.setState({ legacyMode: true })
  }

  openImageDialog () {
    this.refs.checkpointQR.openImageDialog()
  }

  render () {
    const { mode, checkpointKey, checkpointTime, legacyMode } = this.state
    const { urlScanState } = this.props
    const computedMode = typeof urlScanState !== 'undefined'
      ? urlScanState
      : mode
    const qrValue = `${window.location.href}?checkpoint=${checkpointKey}`
    let content
    if (computedMode === 'home') {
      content = (
        <Grid
          container
          direction='column'
          justify='center'
          alignItems='center'
        >
          <Typography style={{ marginTop: 25, marginBottom: 25 }}>
            <Translation>{t => t('welcomeMessage')}</Translation>
          </Typography>
          <Button onClick={this.joinCheckpoint.bind(this)} variant='contained' color='primary' aria-label='add' style={{ marginTop: 50 }}>
            <CropFreeIcon />
            <Translation>{t => t('joinCheckpointButton')}</Translation>
          </Button>
          <Button onClick={this.becomeHost.bind(this)} variant='contained' color='secondary' aria-label='add' style={{ marginTop: 50 }}>
            <HomeIcon />
            <Translation>{t => t('hostCheckpointButton')}</Translation>
          </Button>
        </Grid>
      )
    } else if (computedMode === 'host') {
      content = (
        <Grid
          container
          direction='column'
          justify='center'
          alignItems='center'
        >
          <Typography style={{ marginTop: 25, marginBottom: 25 }}>
            <Translation>{t => t('hostingCheckpointMessage')}</Translation>
          </Typography>
          <QRCode value={qrValue} size={200} style={{ backgroundColor: '#fff', padding: 20 }} />
          <Button onClick={this.endHost.bind(this)} variant='contained' color='primary' aria-label='add' style={{ marginTop: 25 }}>
            <StopIcon />
            <Translation>{t => t('endCheckpointButton')}</Translation>
          </Button>
          <Typography style={{ marginTop: 25 }}>
            <Translation>{t => t('checkpointCreatedMessage')}</Translation> {new Date(checkpointTime).toString()}
          </Typography>
        </Grid>
      )
    } else if (computedMode === 'join') {
      content = (
        <Grid
          container
          direction='column'
          justify='center'
          alignItems='center'
        >
          <QRReader
            ref='checkpointQR'
            delay={300}
            onError={this.handleScanError.bind(this)}
            onScan={this.handleScan.bind(this)}
            style={{ width: legacyMode ? 0 : '100%' }}
            facingMode='environment'
            legacyMode={legacyMode}
          />
          { legacyMode && (
            <Typography style={{ marginTop: 25 }}>
              <Translation>{t => t('noCameraPermissionMessage')}</Translation>
            </Typography>
          ) }
          <Button onClick={this.openImageDialog.bind(this)} variant='contained' color='primary' aria-label='add' style={{ marginTop: 25 }}>
            <Translation>{t => t('takePictureButton')}</Translation>
          </Button>
          <Button onClick={this.reset.bind(this)} variant='contained' color='primary' aria-label='add' style={{ marginTop: 25 }}>
            <ArrowBackIcon />
            <Translation>{t => t('backButton')}</Translation>
          </Button>
        </Grid>
      )
    } else if (computedMode === 'scan-success') {
      content = (
        <Grid
          container
          direction='column'
          justify='center'
          alignItems='center'
        >
          <Typography style={{ marginTop: 25, marginBottom: 25 }}>
            <Translation>{t => t('joinSuccessfulMessage')}</Translation>
          </Typography>
          <Button onClick={this.reset.bind(this)} variant='contained' color='primary' aria-label='add' style={{ marginTop: 25 }}>
            <ArrowBackIcon />
            <Translation>{t => t('backButton')}</Translation>
          </Button>
        </Grid>
      )
    } else if (computedMode === 'scan-error') {
      content = (
        <Grid
          container
          direction='column'
          justify='center'
          alignItems='center'
        >
          <Typography style={{ marginTop: 25, marginBottom: 25 }}>
            <Translation>{t => t('scanErrorMessage')}</Translation>
          </Typography>
          <Button onClick={this.reset.bind(this)} variant='contained' color='primary' aria-label='add' style={{ marginTop: 25 }}>
            <ArrowBackIcon />
            <Translation>{t => t('backButton')}</Translation>
          </Button>
        </Grid>
      )
    }
    return content
  }
}

export default Checkpoints
