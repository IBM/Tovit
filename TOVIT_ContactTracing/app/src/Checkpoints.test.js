/* globals test, expect */
import React from 'react'
import Checkpoints from '../src/Checkpoints'
import TestRenderer from 'react-test-renderer'
import { I18nextProvider } from 'react-i18next'
import i18n from '../src/i18n'

test('Checkpoints component', () => {
  const status = false
  const statusLoaded = true
  const component = TestRenderer.create(
    <I18nextProvider i18n={i18n}>
      <Checkpoints status={status} statusLoaded={statusLoaded} />
    </I18nextProvider>)
  expect(component).toMatchSnapshot()
})
