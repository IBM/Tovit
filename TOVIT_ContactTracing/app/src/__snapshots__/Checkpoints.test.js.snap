// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkpoints component 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
>
  <p
    className="MuiTypography-root MuiTypography-body1"
    style={
      Object {
        "marginBottom": 25,
        "marginTop": 25,
      }
    }
  >
    Welcome to TrackCOVID. To participate in the effort to track the spread of the SARS-COV-2 virus, please scan a checkpoint whenever you interact with others in a way that could transmit the virus.
  </p>
  <button
    aria-label="add"
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "marginTop": 50,
      }
    }
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 5v4h2V5h4V3H5c-1.1 0-2 .9-2 2zm2 10H3v4c0 1.1.9 2 2 2h4v-2H5v-4zm14 4h-4v2h4c1.1 0 2-.9 2-2v-4h-2v4zm0-16h-4v2h4v4h2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
      Scan a Checkpoint
    </span>
  </button>
  <button
    aria-label="add"
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "marginTop": 50,
      }
    }
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
        />
      </svg>
      Host a Checkpoint
    </span>
  </button>
</div>
`;
